src/mapreduce.jl

```diff
@@ -188,20 +181,33 @@ end
 ## reduce ##
 ############
 
-@inline reduce(op, a::StaticArray; dims=:, kw...) = _reduce(op, a, dims, kw.data)
+@inline reduce(op, a::StaticArray; dims = :, init = _InitialValue()) =
+    _reduce(op, a, dims, init)
 
 # disambiguation
 reduce(::typeof(vcat), A::StaticArray{<:Tuple,<:AbstractVecOrMat}) =
     Base._typed_vcat(mapreduce(eltype, promote_type, A), A)
 reduce(::typeof(vcat), A::StaticArray{<:Tuple,<:StaticVecOrMatLike}) =
-    _reduce(vcat, A, :, NamedTuple())
+    _reduce(vcat, A, :, _InitialValue())
 
 reduce(::typeof(hcat), A::StaticArray{<:Tuple,<:AbstractVecOrMat}) =
     Base._typed_hcat(mapreduce(eltype, promote_type, A), A)
 reduce(::typeof(hcat), A::StaticArray{<:Tuple,<:StaticVecOrMatLike}) =
-    _reduce(hcat, A, :, NamedTuple())
+    _reduce(hcat, A, :, _InitialValue())
 
-@inline _reduce(op, a::StaticArray, dims, kw::NamedTuple=NamedTuple()) = _mapreduce(identity, op, dims, kw, Size(a), a)
+@inline _reduce(op, a::StaticArray, dims, init = _InitialValue()) =
+    _mapreduce(identity, op, dims, init, Size(a), a)
+
+################
+## (map)foldl ##
+################
+
+@inline mapfoldl(f, op::R, a::StaticArray; init = _InitialValue()) where {R} =
+    _mapfoldl(f, op, :, init, Size(a), a)
+@inline foldl(op::R, a::StaticArray; init = _InitialValue()) where {R} =
+    _foldl(op, a, :, init)
+@inline _foldl(op::R, a, dims, init = _InitialValue()) where {R} =
+    _mapfoldl(identity, op, dims, init, Size(a), a)
```

Comment:
Do we need the `R` here? It doesn't seem to be used.

Reply:
This is yet another "ritual" that I do just by habit to avoid a bad interaction with the specialization heuristics: https://github.com/JuliaLang/julia/issues/32761#issuecomment-518773597

> If I remember correctly, the `f::F where F` specialization was only required if you didn't call `f` directly inside the function body but instead just passed it to another function — probably a non-inlined function. Or at least that's the way it worked a long time ago (year+).

I thought this can happen when `_foldl` is happened to be at the boundary of inlining decision but I don't have a proof.  If you don't want include this kind of "cargo cult" code, I'm happy to remove the no-op type parameters.

Reply:
Eek, I didn't know about this corner of the specialization heuristics. Would be nice if `@specialize` did this rather than having to introduce `where` syntax.

Anyway, it seems this has been documented, so maybe we should just add a small comment and link to https://docs.julialang.org/en/v1.5-dev/manual/performance-tips/#Be-aware-of-when-Julia-avoids-specializing-1

I'm still a bit unclear on how that relates to inlining. Presumably this has no effect if all callees are inlined, but if `_mapfoldl` was not inlined we'll get hit with dynamic dispatch without this?

Reply:
> if `_mapfoldl` was not inlined we'll get hit with dynamic dispatch without this?

That's my understanding. I thought it's something like the `g_func` example in the documentation.